### ISSUE TITLE
Using the cookbook wrapper method to write out the zone files

### ISSUE THREAD

**Issue by jhammerman79** [joyofhex/cookbook-bind#22]
Hi cookbook-bind maintainers,

I have used your cookbook with a wrapper around it to successfully do everything I want to do with it except actually write out the zone files. As I understand, this is left as something of an exrecise for the reader (at least until the pending merge occurs which incorporates zone file templating.

The issue is that I need to insert custom logic prior to named-checkconf being called; it also has to happen _after_ the directories are created.

I suppose I could pilfer the minimum amount of logic necessary from the default.rb recipe to create the directories, then deliver the files to the correct location on disk, then execute the default recipe in full.

Is there another approach you would recommend? Apologies for creating an issue, but I could not determine an alternative method of contacting you.

Thanks in advance for your help,
Joseph Hammerman

**Comment by jhammerman79** [comment]
Upon reflection, it looks like I will only need to expand the var_cookbook_files attribute to include my zone files.

Thanks!
Joseph Hammerman

**Comment by atomic-penguin** [comment]
I am also on IRC, and noticed you talking with coderanger about the cookbook.  I was busy with a software release for work, and a crisis affecting another cookbook's users yesterday.

If you would not mind, would you please take some time to construct a note describing the problem and solution you found, for this cookbook's README?  It would be a nice thing to have documented procedures for others looking to achieve the same approach.

**Comment by joyofhex** [comment]
I've run into, essentially, the same problem. I'm generating zone files in a wrapper that includes the bind cookbook. This is problematic, because the cookbook creates the directory for zones, populates it with static zone files, writes out a configuration, then used checkconf to verify the config (which includes validating that the zone files exist). There's no extension point where I can add in new zone files without modifying the bind cookbook itself (and I'd rather not maintain what is essentially a fork).

I've resorted for now to create the directory /var/named/master in the wrapper cookbook (and thus have a cloned resource complaint from chef). Whilst it works, it doesn't feel very clean to me.

There's a couple of ways I can think of to solve this:

* extend the zonesource to be able to include recipes from other cookbooks (so that I can have it call a recipe in my wrapper that creates the zone files and adds them to the zone list). Something like:

```ruby
 if !node['bind']['zonesource'].nil?
-  include_recipe "bind::#{node['bind']['zonesource']}2zone"
+  if node['bind']['zonesource'].include? '::'
+    include_recipe node['bind']['zonesource']
+  else
+    include_recipe "bind::#{node['bind']['zonesource']}2zone"
+  end
 else
```

* split the recipe into multiple recipes with fewer responsibilities and just include_recipe them from the default recipe  (so at the least have a recipe for creating the directory structure and then one for everything else. My wrapper can include the individual recipes directly and it can create the appropriate files and populate the zones array appropriately).
* allow the ability to include master zone files from outside of the created hierarchy

**Comment by simoninkin** [comment]
We are having the same issue with this cookbook. In fact, this is not even an issue, more like a design flaw.
Why does this cookbook even try doing the check on files, when they are deffinitely not there?
I mean, whenever you try adding your own zone,  the cookbook would also add the line in the conf with the zonefile location and then run the check againt that, which would fail 100% of the time.

Public Chef Supermarket is missing a cookbook rating system!

**Comment by joyofhex** [comment]
Since my comment all the way back in 2015 I've become the maintainer of the cookbook. I've just released v2.0.0 which by migrating to using custom resources allows the addition of zone files through the use of the appropriate custom resource.
